File: CS/1329_Sort_the_Matrix_Diagonally.cs
public class Solution {
    int[][] result;
    int m;
    int n;
    public int[][] DiagonalSort(int[][] mat) {
        m = mat.Length;
        n = mat[0].Length;
        
        result = mat;
        
        for(int i = 0; i < m; i++){
            sortFunc(i, 0);
        }
        for(int j = 1; j < n; j++){
            sortFunc(0, j);
        }
        
        return result;
    }
    private void sortFunc(int i, int j){
        int length = Math.Min(m-i, n-j);
        int[] temp = new int[length];
        int x, y, k;
        x = i;
        y = j;
        k = 0;
        while(x < m && y < n){
            temp[k] = result[x][y];
            x++;
            y++;
            k++;
        }
        Array.Sort(temp);
        
        x = i;
        y = j;
        k = 0;
        while(x < m && y < n){
            result[x][y] = temp[k];
            x++;
            y++;
            k++;
        }
    }
}

File: Cpp/1329_Sort_the_Matrix_Diagonally.cpp
class Solution {
public:
    int m;
    int n;
    vector<vector<int>> diagonalSort(vector<vector<int>>& mat) {
        
        m = mat.size();
        n = mat[0].size();
        int i = m-2;
        int j = 0;
        
        while(i>=0){
            helper(mat, i, j);
            i--;
        }
        
        i = 0;
        j = 1;
        while(j<n){
            helper(mat, i, j);
            j++;
        }
        
        return mat;
    }
private:
    void helper(vector<vector<int>>& mat, int i , int j){
        vector<int> v;
        int vi = i;
        int vj = j;
        while(vi<m && vj<n){
            v.push_back(mat[vi][vj]);
            vi++;
            vj++;
        }
        sort(v.begin(), v.end());

        vi = i;
        vj = j;
        for(int item : v){
            mat[vi][vj] = item;
            vi++;
            vj++;
        }
    }
};
